torch/nn/utils/stateless.py

```diff
@@ -37,21 +36,83 @@ def _setattr(self, name: str, value: Any) -> None:
-        warnings.warn("functional_call was passed multiple values for tied weights. "
-                      "This behavior is deprecated and will be an error in future versions")
+def _create_tied_weights_map(module: 'torch.nn.Module', params_and_buffers: Dict[str, Tensor]) -> Dict[str, str]:
+    """
+    _create_tied_weights_map(module: Module, params_and_buffers: Dict[str, Tensor]) -> Dict[str, str]
+
+    Creates a weight map of {tied_name: name_given_by_user} for all weights where one of their tied weights is passed
+
+    ex: Foo() has self.foo and self.tied_foo, which are tied. If a user passed {'foo': ...} as the reparamaterization,
+        this would return {'tied_foo': 'foo'}. Similarly if a user passed {'tied_foo': ...}, this returns
+        {'tied_foo': 'foo'}.
+
+    ex: If there aren't any tied weights and the user passed values for every parameter and buffer, this will return a
+        map where every name maps to an empty set: {'l1.weight': set(), 'l1.bias': set(), ...}
 
+    ex: The map only contains values that a user is reparamaterizing. For example, if module = nn.Linear(...) and the
+        user only passed a new value for 'bias', this looks returns: {'bias': set()}
 
-def _create_swap_params(params_and_buffers, replaced_tensors_map):
-    def _swap_parameters(module, tensor_name: str, full_path: str, tensor: Tensor) -> None:
+    This is useful because we will start by reparamaterizing all the keys of params_and_buffers, then all the key from
+    this returned dictionary.
+    """
+
+    # The basic algorithm looks like:
+    #   - index all weights by their original tensor value to find tied weights
+    #     - when we encounter a weight not used by the user, we save it in a set (second element in the tuple)
+    #     - when we run into a weight used by the user, we save that separate from the set as the first element in the tuple
+    #     - ending map looks like {tensor: (name_given_by_user, set(all_tied_names)}
+    #   - then loop through the values of this map (name_given_by_user and set(all_tied_names))
+    #     - for each element of all_tied_names, add {tied_name: name_given_by_user} to a new map
+
+    names = params_and_buffers.keys()
+    weight_to_name_and_tied_names: Dict[torch.Tensor, Tuple[Optional[str], Set[str]]] = {}
+
+    # create a map keyed by tensor value so that tied weights get mapped to the same key. The value is the interesting
+    # part at the end it's (used_name, (tied_names)).
+    # For example, in the first example where there's tied weights self.foo and self.tied_foo and the user passes a
+    # value for self.foo, this will return {torch.Tensor(...): ('foo', set('tied_foo'))}
+    def add_to_name_map(n: str, t: torch.Tensor):
+        # if the tensor hasn't been seen before, add it to the map
+        if t not in weight_to_name_and_tied_names:
+            weight_to_name_and_tied_names[t] = (n, set()) if n in names else (None, {n})
```

Comment:
for `n in names`, I am not sure what exact object the `dict_keys` object is. What is the complexity of this lookup?

Reply:
O(1) according to the internet: https://stackoverflow.com/questions/17539367/python-dictionary-keys-in-complexity